File: Day 23 - PHP/solution.py
from collections import defaultdict

import sys
mem = defaultdict(int)
instructions = [x.strip().split() for x in sys.stdin.readlines()]

ip = 0
mulled = 0

while ip >= 0 and ip < len(instructions):
    instruction= instructions[ip]
    action, a, b = instruction
    a_val = None
    b_val = None
    ipjump = 1

    if a.islower():
        a_val = mem[a]
    else:
        a_val = int(a)
    if b.islower():
        b_val = mem[b]
    else:
        b_val = int(b)

    if action == 'set':
        mem[a] = b_val
    elif action == 'sub':
        mem[a] -= b_val
    elif action == 'mul':
        mulled += 1
        mem[a] *= b_val
    elif action == 'jnz':
        if a_val:
            ipjump = b_val

    ip += ipjump

    
print("Part 1:", mulled)

b, h = 107900, 0
for x in range(b, b + 17000 + 1, 17):
    for i in range(2, x):
        if x % i == 0:
            h += 1
            break

print("Part 2:", h)

File: Day 23 - PHP/solution.php
<?php
$registers = [];
$get = function ($register) use (&$registers) {
    if (array_key_exists($register, $registers)) {
        return $registers[$register];
    } else {
        return 0;
    }
};

$instructions = [];
while ($line = fgets(STDIN)) {
    $instructions[] = rtrim($line);
}

$ip = 0;
$mulled = 0;
while ($ip < count($instructions)) {
    list($action, $a, $b) = explode(" ", $instructions[$ip]);
    $ipjump = 1;

    $a_val = $a;
    if (ctype_alpha($a)) {
        $a_val = $get($a);
    }

    $b_val = $b;
    if (ctype_alpha($b)) {
        $b_val = $get($b);
    }

    if ($action === "set") {
        $registers[$a] = $b_val;
    } else if ($action === "sub") {
        $registers[$a]-= $b_val;
    } else if ($action === "mul") {
        $mulled+= 1;
        $registers[$a]*= $b_val;
    } else if ($action === "jnz") {
        if ($a_val !== 0) {
            $ipjump = $b_val;
        }
    }
    $ip+= $ipjump;
}

echo "Part 1: $mulled\n";

$b = 107900;
$h = 0;
for ($x = $b;$x <= ($b + 17000);$x+= 17) {
    for ($e = 2;$e < $x;$e++) {
        if ($x % $e == 0) {
            $h++;
            break;
        }
    }
}

echo "Part 2: $h\n";
?>
